Multithreading option not working

The code below works fine if `acceleration` option is omitted.

```julia
using MLJ # or `MLJBase, MLJModels, MLJEnsembles` for minimal install

DecisionTreeClassifier = @iload DecisionTreeClassifier pkg=DecisionTree

atom = DecisionTreeClassifier()
model = EnsembleModel(
    atom;
    bagging_fraction=0.6,
    rng=123,
    acceleration=CPUThreads(),
    out_of_bag_measure = [log_loss, accuracy],
)

X, y = @load_iris # a table and a vector

julia> mach = machine(model, X, y) |> fit!
[ Info: Training machine(ProbabilisticEnsembleModel(model = DecisionTreeClassifier(max_depth = -1, …), …), …).
Ensemble-building in parallel on 5 threads.
┌ Error: Problem fitting the machine machine(ProbabilisticEnsembleModel(model = DecisionTreeClassifier(max_depth = -1, …), …), …). 
└ @ MLJBase ~/MLJ/MLJBase/src/machines.jl:682
[ Info: Running type checks... 
[ Info: Type checks okay. 
ERROR: TaskFailedException
Stacktrace:
  [1] wait
    @ ./task.jl:345 [inlined]
  [2] threading_run(fun::MLJEnsembles.var"#170#threadsfor_fun#29"{MLJEnsembles.var"#170#threadsfor_fun#28#30"{typeof(MLJEnsembles.get_ensemble_and_indices), Vector{Any}, Int64, Int64, Tuple{Matrix{Float64}, Vector{UInt32}, Vector{Symbol}, CategoricalArrays.CategoricalVector{String, UInt32, String, CategoricalArrays.CategoricalValue{String, UInt32}, Union{}}}, ProgressMeter.Progress, Random.MersenneTwister, Int64, Int64, MLJDecisionTreeInterface.DecisionTreeClassifier, UnitRange{Int64}}}, static::Bool)
    @ Base.Threads ./threadingconstructs.jl:38
  [3] macro expansion
    @ ./threadingconstructs.jl:89 [inlined]
  [4] _fit(res::CPUThreads{Nothing}, func::Function, verbosity::Int64, stuff::Tuple{MLJDecisionTreeInterface.DecisionTreeClassifier, Int64, Int64, Int64, Random.MersenneTwister, ProgressMeter.Progress, Tuple{Matrix{Float64}, Vector{UInt32}, Vector{Symbol}, CategoricalArrays.CategoricalVector{String, UInt32, String, CategoricalArrays.CategoricalValue{String, UInt32}, Union{}}}})                                                                               
    @ MLJEnsembles ~/.julia/packages/MLJEnsembles/NtgQL/src/ensembles.jl:414
  [5] fit(::MLJEnsembles.ProbabilisticEnsembleModel{MLJDecisionTreeInterface.DecisionTreeClassifier}, ::Int64, ::NamedTuple{(:sepal_length, :sepal_width, :petal_length, :petal_width), NTuple{4, Vector{Float64}}}, ::CategoricalArrays.CategoricalVector{String, UInt32, String, CategoricalArrays.CategoricalValue{String, UInt32}, Union{}})      
    @ MLJEnsembles ~/.julia/packages/MLJEnsembles/NtgQL/src/ensembles.jl:476
  [6] fit_only!(mach::Machine{MLJEnsembles.ProbabilisticEnsembleModel{MLJDecisionTreeInterface.DecisionTreeClassifier}, true}; rows::Nothing, verbosity::Int64, force::Bool, composite::
Nothing)                                                                                   
    @ MLJBase ~/MLJ/MLJBase/src/machines.jl:680
  [7] fit_only!
    @ ~/MLJ/MLJBase/src/machines.jl:606 [inlined]
  [8] #fit!#63
    @ ~/MLJ/MLJBase/src/machines.jl:778 [inlined]
  [9] fit!
    @ ~/MLJ/MLJBase/src/machines.jl:775 [inlined]
 [10] |>(x::Machine{MLJEnsembles.ProbabilisticEnsembleModel{MLJDecisionTreeInterface.DecisionTreeClassifier}, true}, f::typeof(fit!))
    @ Base ./operators.jl:911
 [11] top-level scope
    @ REPL[31]:1

    nested task error: AssertionError: length(ints) == 501
    Stacktrace:
      [1] mt_setfull!(r::Random.MersenneTwister, #unused#::Type{UInt64})
        @ Random /Applications/Julia-1.8.app/Contents/Resources/julia/share/julia/stdlib/v1.8/Random/src/RNGs.jl:226
      [2] reserve1
        @ /Applications/Julia-1.8.app/Contents/Resources/julia/share/julia/stdlib/v1.8/Random/src/RNGs.jl:257 [inlined]
      [3] mt_pop!
        @ /Applications/Julia-1.8.app/Contents/Resources/julia/share/julia/stdlib/v1.8/Random/src/RNGs.jl:262 [inlined]
      [4] rand
        @ /Applications/Julia-1.8.app/Contents/Resources/julia/share/julia/stdlib/v1.8/Random/src/RNGs.jl:419 [inlined]
      [5] rand
        @ /Applications/Julia-1.8.app/Contents/Resources/julia/share/julia/stdlib/v1.8/Random/src/Random.jl:257 [inlined]
      [6] rand(rng::Random.MersenneTwister, sp::Random.SamplerRangeNDL{UInt64, Int64})
        @ Random /Applications/Julia-1.8.app/Contents/Resources/julia/share/julia/stdlib/v1.8/Random/src/generation.jl:344
      [7] rand
        @ /Applications/Julia-1.8.app/Contents/Resources/julia/share/julia/stdlib/v1.8/Random/src/Random.jl:254 [inlined]
      [8] fisher_yates_sample!(rng::Random.MersenneTwister, a::UnitRange{Int64}, x::Vector{Int64})                                                                                    
        @ StatsBase ~/.julia/packages/StatsBase/XgjIN/src/sampling.jl:199
      [9] sample!(rng::Random.MersenneTwister, a::UnitRange{Int64}, x::Vector{Int64}; replace::Bool, ordered::Bool)                                                               
        @ StatsBase ~/.julia/packages/StatsBase/XgjIN/src/sampling.jl:484
     [10] #sample#207
        @ ~/.julia/packages/StatsBase/XgjIN/src/sampling.jl:511 [inlined]
     [11] #15
        @ ./none:0 [inlined]
     [12] iterate
        @ ./generator.jl:47 [inlined]
     [13] collect_to!
        @ ./array.jl:845 [inlined]
     [14] collect_to_with_first!
        @ ./array.jl:823 [inlined]
     [15] collect(itr::Base.Generator{UnitRange{Int64}, MLJEnsembles.var"#15#17"{Int64, Int64, Random.MersenneTwister}})         
        @ Base ./array.jl:797
     [16] get_ensemble_and_indices
        @ ~/.julia/packages/MLJEnsembles/NtgQL/src/ensembles.jl:153 [inlined]
     [17] macro expansion
        @ ~/.julia/packages/MLJEnsembles/NtgQL/src/ensembles.jl:416 [inlined]
     [18] (::MLJEnsembles.var"#170#threadsfor_fun#29"{MLJEnsembles.var"#170#threadsfor_fun#28#30"{typeof(MLJEnsembles.get_ensemble_and_indices), Vector{Any}, Int64, Int64, Tuple{Matrix{Float64}, Vector{UInt32}, Vector{Symbol}, CategoricalArrays.CategoricalVector{String, UInt32, String, CategoricalArrays.CategoricalValue{String, UInt32}, Union{}}}, ProgressMeter.Progress, Random.MersenneTwister, Int64, Int64, MLJDecisionTreeInterface.DecisionTreeClassifier, UnitRange{Int64}}})(tid::Int64; onethread::Bool)
        @ MLJEnsembles ./threadingconstructs.jl:84
     [19] #170#threadsfor_fun
        @ ./threadingconstructs.jl:51 [inlined]
     [20] (::Base.Threads.var"#1#2"{MLJEnsembles.var"#170#threadsfor_fun#29"{MLJEnsembles.var"#170#threadsfor_fun#28#30"{typeof(MLJEnsembles.get_ensemble_and_indices), Vector{Any}, Int64, Int64, Tuple{Matrix{Float64}, Vector{UInt32}, Vector{Symbol}, CategoricalArrays.CategoricalVector{String, UInt32, String, CategoricalArrays.CategoricalValue{String, UInt32}, Union{}}}, ProgressMeter.Progress, Random.MersenneTwister, Int64, Int64, MLJDecisionTreeInterface.DecisionTreeClassifier, UnitRange{Int64}}}, Int64})()
        @ Base.Threads ./threadingconstructs.jl:30
```

--- a/src/ensembles.jl
+++ b/src/ensembles.jl
@@ -408,31 +408,56 @@ function _fit(res::CPUProcesses, func, verbosity, stuff)
     end
 end
 
+# Create thread safe version of RNGs.
+# Random._GLOBAL_RNG() and Random.default_rng() are threadsafe by default_rng
+# as they have thread local state from julia >=1.3<=1.6 and task local state Julia >=1.7   
+threadsafe_rng(rng::typeof(Random.default_rng())) = rng
+threadsafe_rng(rng::Random._GLOBAL_RNG) = rng
+threadsafe_rng(rng) = deepcopy(rng)
+
 function _fit(res::CPUThreads, func, verbosity, stuff)
     atom, n, n_patterns, n_train, rng, progress_meter, args = stuff
     if verbosity > 0
         println("Ensemble-building in parallel on $(Threads.nthreads()) threads.")
     end
+    
     nthreads = Threads.nthreads()
+    
+    if nthreads == 1
+        return  _fit(CPU1(), func, verbosity, stuff)
+    end
+
     chunk_size = div(n, nthreads)
     left_over = mod(n, nthreads)
     resvec = Vector(undef, nthreads) # FIXME: Make this type-stable?
 
-    Threads.@threads for i = 1:nthreads
-        resvec[i] = if i != nworkers()
-            func(atom, 0, chunk_size, n_patterns, n_train, rng, progress_meter, args...)
-        else
-            func(
-                atom,
-                0,
-                chunk_size + left_over,
-                n_patterns,
-                n_train,
-                rng,
-                progress_meter,
-                args...,
+    @sync begin
+        for i in 1:nthreads-1
+            Threads.@spawn(
+                resvec[i] = func(
+                    atom,
+                    0,
+                    chunk_size,
+                    n_patterns,
+                    n_train,
+                    threadsafe_rng(rng),
+                    progress_meter,
+                    args...
+                )
             )
         end
+        Threads.@spawn(
+            resvec[nthreads] = func(
+                    atom,
+                    0,
+                    chunk_size + left_over,
+                    n_patterns,
+                    n_train,
+                    threadsafe_rng(rng),
+                    progress_meter,
+                    args...
+            )
+        )
     end
 
     return reduce(_reducer, resvec)

--- a/test/ensembles.jl
+++ b/test/ensembles.jl
@@ -257,7 +257,7 @@ end
 
     @test !isnan(predict(ensemble, MLJEnsembles.selectrows(X, test))[1])
 
-    # tests using integer rngs (see issue 223)
+    # tests using integer rngs (see issue 27)
     X_, y_ = @load_iris
     atom = KNNClassifier(K = 7)
     ensemble_model = EnsembleModel(

--- a/test/ensembles.jl
+++ b/test/ensembles.jl
@@ -256,7 +256,22 @@ end
     @test length(ensemble.fitresult.ensemble) == 5
 
     @test !isnan(predict(ensemble, MLJEnsembles.selectrows(X, test))[1])
+
+    # tests using integer rngs (see issue 223)
+    X_, y_ = @load_iris
+    atom = KNNClassifier(K = 7)
+    ensemble_model = EnsembleModel(
+        atom;
+        bagging_fraction=0.6,
+        rng=123,
+        out_of_bag_measure = [log_loss, brier_score]
+    )
+    ensemble = machine(ensemble_model, X_, y_)
+    fit!(ensemble)
+    @test length(ensemble.fitresult.ensemble) == ensemble_model.n
+
 end
+
 
 end
 

